LX-309 Modify CAPA choicegroup problems to make custom radio/checkboxes possible

pomegranited opened the pull request:
The LabXchange project requires custom styling of the radio/checkbox problems, which, due to the element structure, could not be done with CSS alone.  We made some changes to the DOM to facilitate these customizations, which this PR submits upstream:

* Move the `<input>` items out of their `<label>`, to be siblings, as [is already done in Paragon](https://github.com/edx/paragon/blob/01861049f8f75fec0b0f7d0dbc40e8def58938e1/.storybook/__snapshots__/Storyshots.test.js.snap#L3293-L3313).
  This is required so that CSS selectors like these can be used which target the dynamic attributes of the input element:
  ```
  input:hover + label,
  input:checked + label 
  ```
  This is not possible without using sibling selectors, until [browsers begin supporting `:has`](https://caniuse.com/#search=%3Ahas) yet.  See [stackoverflow thread](https://stackoverflow.com/questions/45031573/apply-style-to-wrapping-label-of-checked-checkbox) for more info.
* Move the `.status` icon out of the radio button label.
  This makes the status display more consistent with the checkbox multiple choice items (see screenshot below).
* Adds a `submitted` class to mark radio/checkbox items which were submitted as answers during the previous attempt.
  This allows us to style these options differently in LabXchange.

**JIRA tickets**:  OSPR

**Screenshots**:

Status icon for radio button problems moved:

![moves status icon](https://user-images.githubusercontent.com/7556571/61857410-44782d00-af03-11e9-8965-5b6e3bf0a880.png)

The other UI changes do not change the appearance of the CAPA problems.

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:

TBD

**Reviewers**
- [ ] @symbolist 
- [ ] edX reviewer[s] TBD

CC @ormsbee

edx-webhook commented:
Thanks for the pull request, @pomegranited! I've created [OSPR-3760](https://openedx.atlassian.net/browse/OSPR-3760) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

natabene commented:
@pomegranited Thank you for your contribution. Please let me know once it is ready for our review.

pomegranited commented:
Thanks @natabene -- I still need to get the tests passing again, but could you run this change past the a11y team (e.g. @wittjeff ) to ensure it meets their accessibility standards?

natabene commented:
@wittjeff Can you give this a look as soon as you can?

wittjeff commented:
Hi @pomegranited !  What you have submitted looks OK but there are a few additional things to consider:
1) There is a WCAG 2.0 accessibility guideline that warns against use of color alone to indicate state. https://www.w3.org/TR/UNDERSTANDING-WCAG20/visual-audio-contrast-without-color.html . So by that criterion, I think there needs to be a word somewhere in the block saying "submitted" (visible to all). I'm not sure where in the UI would be ideal though.
1a) I wonder if it might make sense to change the action button text to Resubmit once the selection state has changed (along with the removal of "submitted" state).
1b) I'm of the opinion that it would be better to have text/graphic redundancy with the Correct (check) and Incorrect (X) graphics.  Further I think it might make most sense to combine the state info there. So perhaps it would say "Submitted: [check] Correct".  Under this scheme the checkmark and X should have alt="" to avoid text output repetition.
2) Incidentally, I have been reviewing the platform for WCGA 2.1 conformance and there are some related changes that I would like to suggest.  Maybe you would like to pick them up now? Or they could be folded in later of course.
a) I think it would be good if the radio and checkbox sizes were a bit bigger, for better visual noticeability.  Perhaps 22x22px?  
b) WCAG 2.1 requires a luminance contrast ratio of 3:1 or better for the borders of active elements. So #949494 would be better for the borders of the radio and checkbox inputs.
c) WCAG 2.1 also has a 3:1 minimum luminance contrast requirement for state info.  I suggest adding a 1px box-shadow (in all 4 directions) to the border, instead of the browser default, for visual consistency across browsers.
d) Related to c), it would be ideal if i) the grey used for "selected"+"submitted", and ii) the blue used for "selected"+"unsubmitted",  each had minimum 3:1 contrast vs the (presumed white) background, AND vs. each other.  This might require a darker grey (#949494) and a darker blue.  I think the new blue being used in edx.org discovery pages may work for this.  But I need to confirm the intended scope of use for that blue.
If the items in 2) are confusing I can make a demo page just to illustrate.
Thanks!

pomegranited commented:
Awesome, thank you for your reply @wittjeff !  I was mainly worried about the `<input><label>` change, but glad it won't be a problem.  Will see how many of your suggestions I can get through here too, and will tackle at least #1, if not most of #2.  Of the changes under #2, are any higher priority than the others?

pomegranited commented:
@wittjeff 

> 1a) I wonder if it might make sense to change the action button text to Resubmit once the selection state has changed (along with the removal of "submitted" state).

Totally agree.  Added with 4f5836d.

> 1b) I'm of the opinion that it would be better to have text/graphic redundancy with the Correct (check) and Incorrect (X) graphics. Further I think it might make most sense to combine the state info there. So perhaps it would say "Submitted: [check] Correct". 

I agree that we should be showing "Correct" or "Incorrect" text along with the status icon -- the icon by itself is ambiguous.  But I think that adding "Submitted:" starts making it pretty wordy.  I've done as you requested though, so you can have a look.  What do you think, is it ok to drop the "Submitted:" ? 

![Submitted (x) incorrect](https://user-images.githubusercontent.com/7556571/62368327-732d7d80-b56b-11e9-8640-22a543e9dc76.png)

> Under this scheme the checkmark and X should have alt="" to avoid text output repetition.

Currently, the text that's shown on hover is a "tooltip", not an "alt" text, and they read rather nicely:
* This answer is correct.
* This answer is incorrect.

Do you want these tooltips removed?

> 2a) I think it would be good if the radio and checkbox sizes were a bit bigger, for better visual noticeability. Perhaps 22x22px?

Done with 47d30ef -- see screenshot above.

> If the items in 2) are confusing I can make a demo page just to illustrate.

Could you?  I'm getting confused about which borders you mean, and want to make sure I get the colours right.  Thank you!

pomegranited commented:
This is ready for your review @symbolist .

Just waiting on a couple of things before I can complete @wittjeff 's [requests above](https://github.com/edx/edx-platform/pull/21215#issuecomment-515595848):
* Should I remove the "tooltips" from the status text?
* Demo page requested to demonstrate colors/borders for item 2.

symbolist commented on common/lib/xmodule/xmodule/js/spec/capa/display_spec.js:
```diff
@@ -365,13 +365,19 @@ data-url='/problem/quiz/'> \
-      expect(this.problem.submitButtonLabel.text()).toBe('Submit');
-      expect(this.problem.submitButton).not.toHaveAttr('disabled');
+
+     ['Submit', 'Resubmit'].forEach(function(button_label) {
```
`button_label` is only being used on line 375. Maybe doing a second submit and checking that the label still remains "Resubmit" would be better?

symbolist commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -15,47 +15,39 @@
-                ${describedby_html}
-                >
-                <input type="${input_type}" name="input_${id}${name_array_suffix}" id="input_${id}_${choice_id}" class="field-input input-${input_type}" value="${choice_id}"
+            if is_radio_input(choice_id) or input_type != 'radio' and choice_id in value:
```
Can you add brackets to make this clearer?

symbolist commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -15,47 +15,39 @@
+          %>
+          <div class="field">
+            <input type="${input_type}" name="input_${id}${name_array_suffix}" id="input_${id}_${choice_id}"
+              class="${input_class}" value="${choice_id}"
```
Is the change in the `class` value intentional?

symbolist replied:
Got it.

symbolist commented on common/lib/capa/capa/templates/choicegroup.html:
```diff
@@ -15,47 +15,39 @@
         <span id="answer_${id}"></span>
     </fieldset>
     <div class="indicator-container">
-        % if input_type == 'checkbox' or status.classname == 'unanswered':
```
Can you explain why we don't need this?

symbolist replied:
Ah okay.

pomegranited commented:
@natabene  This is ready for edX review.

[As noted above](https://github.com/edx/edx-platform/pull/21215#issuecomment-518346728), I still have a couple of outstanding changes requested by @wittjeff , but am waiting on his reply there.

natabene commented:
@marcotuts Do you want to have a look before this goes to engineering?

wittjeff commented:
1) Yes I think you can remove the tooltips but I don't think it is urgent.
2) Thanks for the icon size nudge.  Re the colors, I have Arbisoft people working on this and some related ones now so there is no need to bother with it now. 
Thanks!

pomegranited commented:
Thanks @wittjeff !  I went ahead and removed the tooltips.

pomegranited commented:
Hi @marcore  CC @natabene -- have you had a chance to look at these changes?

marcotuts commented:
@wittjeff  - can we connect about these changes? I have some concerns about the introduced text here and would like to find an alternate path that does not introduce this.

wittjeff commented:
@marcotuts The switch to symbol+text is not a strict accessibility requirement though I think it is stronger for understanding, so really a usability thing. Is there a problem with i18n?

22px has been suggested as a minimum touch target size in other contexts so that was just a nudge in that direction (not strictly needed here because the whole input+label region is active, but made with future consistency in mind). 

I see your meeting request, can wait til then to discuss if you'd prefer.

ormsbee commented:
@pomegranited: Nice! I'm starting code review now, but it looks like the sandbox isn't working. Is it possible to get it restarted?

As is always the case when we're playing in Capa-land, my main concern is what this means for compatibility with existing content. Are there corresponding Studio docs that need to be updated?

Thank you!

FYI @pdpinch, @Colin-Fredericks

pomegranited commented:
Thanks @ormsbee !

> it looks like the sandbox isn't working. Is it possible to get it restarted?

Yep, I've built a new one, so those links should work now.

> As is always the case when we're playing in Capa-land, my main concern is what this means for compatibility with existing content. 

The sandbox has the standard DemoX course on it, so you can see that it works with old imported problems.

> Are there corresponding Studio docs that need to be updated?

There's no XML changes here, just GUI, so there may be some out-of-date screenshots in the docs?  But no change to the author instructions or anything like that.

Colin-Fredericks commented:
I am way too swamped to test this week.

If you'd like to try out my usual set of courses, here's the [Tarballs for Testing](https://drive.google.com/open?id=1UjEnA4q9EpPHyIsw7ZzWAdDGVLDKaB97) folder. I would especially check the more unusual assessment tricks in Grape Ape and the "better basic problems" bit in Studio Advanced.

pomegranited commented:
Hi @natabene @ormsbee  Do you know when we might get this reviewed?  Please let me know if I should rebase on latest master and deploy a new sandbox for testing.  Thanks!

marcotuts commented:
Hello! It looks like my comment was never submitted here after @wittjeff  and I met. :( 
My apologies for that.

For now, the summary I would give is that we should revert the visual changes identified in the comment above: https://github.com/edx/edx-platform/pull/21215#issuecomment-515595848

(For example, we should not add / show the additional "Correct" / "Incorrect" text to the right of the per-input option feedback check marks / x marks.)  In other words, there should be no visual change in this PR relating to this part of the CAPA experience other than behind the scenes div / layout changes necessary to support styling. 

If this summary doesn't make the list of required change reversions clear let me know! If it makes things easier I am happy to jump on a call / time block to discuss this and make sure we speed up response times for this change.

pomegranited commented:
@marcotuts I've removed the "Correct" / "Incorrect" text added here, so the only remaining visual changes are:
* Status icon for radio problems is moved down below the problem, to the same place it is shown for checkbox problems.
* Radio buttons and checkbox problems are slightly larger, 22x22px.
* "Submit" button changes to "Resubmit" when attempts > 0.

I've provisioned a fresh sandbox, so you can preview these changes.  The [Capa Demo course](https://pr21215.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA+2019_1/courseware) has samples of all the permutations I could think of -- show answer, reset, show/hide correctness -- and has the usual demo accounts:

https://pr21215.sandbox.opencraft.hosting

@ormsbee  Have rebased on latest master and squashed down to a single commit.

We have some LX frontend changes that are hinging on this, and so the timing is a little tricky.  We'd really appreciate if this could be merged and deployed to stage.edx.org ASAP, and if all goes ok there, to edx.org before Monday's LX launch.  Do you think that's possible?

CC @symbolist

marcotuts commented:
Hi Jill thanks for the details on what has changed, I left a few thoughts below:

1.) Status icon for radio problems is moved down below the problem, to the same place it is shown for checkbox problems.
So that I make sure I'm looking at the right thing, you are saying that the X icon here is moved down, where it was previously shown in-context within the input element: 
![image](https://user-images.githubusercontent.com/2023680/64622488-6c3d2a80-d3b5-11e9-9434-e0b488095675.png)

If that is the case I think this is fine, though I'm not actually sure if this is necessarily a step forward here? I'm curious if in-context is actually better than the messages below, and that the status below is best used for when there is both in-context and overall scores (especially if the status is different as with partial credit cases). I'm ok with this change though, no action necessary just calling out my thought process on this one. 



2.) Radio buttons and checkbox problems are slightly larger, 22x22px.
This is also ok I think, though generally I wouldn't override defaults and would rely on the same size of the text to the right of the input? No changes needed here. 

3.) "Submit" button changes to "Resubmit" when attempts > 0.
This should not change I believe as part of this work. We should keep the current behavior that shows you the attempts to the right of the button with the Submit language.

ormsbee commented:
I'm re-reviewing after each code commit to try to save roundtrip time. From the code perspective, this is still good. @marcotuts: So as long as the "Resubmit" text is change, we're good to merge?

pomegranited commented:
@marcotuts CC @ormsbee 

> 1.) Status icon for radio problems is moved down below the problem, to the same place it is shown for checkbox problems.
> So that I make sure I'm looking at the right thing, you are saying that the X icon here is moved down, where it was previously shown in-context within the input element:

Yes, that's the change made here.

> I'm curious if in-context is actually better than the messages below, and that the status below is best used for when there is both in-context and overall scores (especially if the status is different as with partial credit cases).

The aim here was for consistency, and we couldn't change the checkbox problems to use in-context status markers without exposing partial correctness of those answers, which would dramatically change the way learners interact with checkbox problems.  (E.g., if I have 2 attempts available, I'd tick all the options first, and then it shows me which ones are correct.)

The per-problem status area is the same even for multi-problem partial credit cases -- partial correctness is shown at the very bottom (see [Show Correctness > Checkbox & Radio & Numeric](https://pr21215.sandbox.opencraft.hosting/courses/course-v1:OpenCraft+CAPA+2019_1/courseware/cd889084c8f54af5990916cf05c80b82/20821a2515584d54a447421105c81bc1/?child=last) for a working example):

![partial correctness](https://user-images.githubusercontent.com/7556571/64667224-4a68a400-d498-11e9-8d00-8482bf02aae7.png)

> 3.) "Submit" button changes to "Resubmit" when attempts > 0.
> This should not change I believe as part of this work. We should keep the current behavior that shows you the attempts to the right of the button with the Submit language.

I've reverted this change with 4acf40d , [see diff](https://github.com/edx/edx-platform/compare/877d534abb14dbef898a8d841061992da995d245..4acf40d61ffa4b834cd8ac982dcad5ee8c49e2f8).  (Force-pushed to keep this PR in a one-commit merge-able state).

The sandbox is also updated, so you can verify the change.

marcotuts commented:
Thank you so much @pomegranited  for the changes here! I can commit to having an internal T&L review / check on the items you backed out of this PR to see what it would take to reconsider items such as the resubmit button text, captured in the diff above. This looks good to go from my standpoint!

edx-webhook commented:
@pomegranited 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/21215&entry.1484655318=https://github.com/pomegranited&entry.752974735=2019-07-25+08:11&entry.1917517419=2019-09-11+16:26&entry.2133058324=Yes) so we can improve your experience in the future.

pomegranited commented:
Thanks for your help and advice here @wittjeff @marcotuts @ormsbee .  
From my perspective, I thought that discussing these changes with @wittjeff  as the a11y expert would be the best way to ensure we weren't going the wrong way, and we took a little time to apply his recommendations.  So it was pretty confusing and counter-productive to have to back them out.  We'd appreciate guidance in the future on how best to get UI changes like these through.
